index/segment/mem/segment.go

```diff
@@ -22,224 +22,246 @@ package mem
-// postings IDs at offset+1.
+// postings IDs at the provided offset.
 func NewSegment(offset postings.ID, opts Options) (sgmt.MutableSegment, error) {
 	s := &segment{
-		opts:      opts,
-		offset:    int(offset) + 1, // The first ID assigned by the segment is offset + 1.
+		offset:    int(offset),
 		newUUIDFn: opts.NewUUIDFn(),
 		termsDict: newTermsDict(opts),
+		readerID:  postings.NewAtomicID(offset),
 	}
 
 	s.docs.data = make([]doc.Document, opts.InitialCapacity())
 
-	s.ids.writer = postings.NewAtomicID(offset)
-	s.ids.reader = postings.NewAtomicID(offset)
+	s.writer.nextID = offset
 	return s, nil
 }
 
-func (s *segment) Insert(d doc.Document) error {
+func (s *segment) Insert(d doc.Document) ([]byte, error) {
 	s.state.RLock()
 	defer s.state.RUnlock()
 	if s.state.closed {
-		return sgmt.ErrClosed
+		return nil, sgmt.ErrClosed
 	}
 
-	if s.state.sealed {
-		return errSegmentSealed
+	ds := []doc.Document{d}
+	if err := s.prepareDocs(ds); err != nil {
+		return nil, err
 	}
 
-	docID, err := d.Validate()
-	if err != nil {
-		return err
-	}
+	// Update the document in case we generated a UUID for it.
+	d = ds[0]
 
-	if docID != nil {
-		ok, err := s.termsDict.ContainsTerm(doc.IDReservedFieldName, docID)
-		if err != nil {
-			return fmt.Errorf("could not lookup document ID '%s'", docID)
-		}
-		if ok {
-			// Index already contains this document.
-			return nil
-		}
-	} else {
-		uuid, err := s.newUUIDFn()
-		if err != nil {
-			return err
-		}
-		d.Fields = append(d.Fields, doc.Field{
-			Name:  doc.IDReservedFieldName,
-			Value: uuid,
-		})
-	}
-
-	// TODO: Consider supporting concurrent writes by relaxing the requirement that
-	// inserted documents are immediately searchable.
-	s.ids.Lock()
+	{
+		s.writer.Lock()
 
-	newID := s.ids.writer.Inc()
-	s.insertDoc(newID, d)
-	err = s.insertTerms(newID, d)
-	s.ids.reader.Inc()
+		s.insertDocWithLocks(d)
+		s.readerID.Inc()
 
-	s.ids.Unlock()
+		s.writer.Unlock()
+	}
 
-	return err
+	return d.ID, nil
 }
 
-func (s *segment) Reader() (index.Reader, error) {
+func (s *segment) Batch(ds []doc.Document) error {
 	s.state.RLock()
+	defer s.state.RUnlock()
 	if s.state.closed {
-		s.state.RUnlock()
-		return nil, sgmt.ErrClosed
+		return sgmt.ErrClosed
 	}
 
-	maxID := s.ids.reader.Load()
-	r := newReader(s, maxID)
-	return r, nil
-}
+	if err := s.prepareDocs(ds); err != nil {
+		return err
+	}
 
-func (s *segment) Seal() error {
-	s.state.Lock()
-	if s.state.sealed {
-		s.state.Unlock()
-		return errSegmentSealed
+	{
+		s.writer.Lock()
+
+		for _, d := range ds {
+			s.insertDocWithLocks(d)
+		}
+		s.readerID.Add(uint32(len(ds)))
+
+		s.writer.Unlock()
 	}
 
-	s.state.sealed = true
-	s.state.Unlock()
 	return nil
 }
 
-func (s *segment) Close() error {
-	s.state.Lock()
-	if s.state.closed {
-		s.state.Unlock()
-		return sgmt.ErrClosed
+func (s *segment) prepareDocs(ds []doc.Document) error {
+	ids := idsgen.New(len(ds))
```

Comment:
Yea, I wanted to avoid this and push it into the caller but I thought that could break some fundamental assumptions if the client wasn't careful. I also considered moving the `prepareDocs` function into the writer lock so it could be reused. How does that sound as an alternative? I definitely think we're going to need to come back and optimize the segment.